Feature/sub pathfinding

RichardKarlFuchs opened the pull request:
Autos erhalten beim spawnen eine Liste an ITracks, die aus der RoutingTable mit einer gewissen Wahrscheinlichkeit ausgewählt wurde. Die Wahrscheinlichkeit wird durch ein RouteProbability-Objekt kontrolliert. Die RoutingTable besteht aus RouteSegments, die aus einer Edge und einer Lane bestehen.
closes #51

RichardKarlFuchs committed:
add method to get required LaneType to turn to next edge

RichardKarlFuchs committed:
add RouteSegment struct and ITrack interface

RichardKarlFuchs committed:
set edge to implement ITrack interface, move GetAbsolutePosition to Edge

RichardKarlFuchs committed:
add RouteSegment Constructor

RichardKarlFuchs committed:
add SubRoute inheritance from IVertex Interface

RichardKarlFuchs committed:
add zip method for three lists

RichardKarlFuchs committed:
construct routing table with RouteSegments instead of Edges

RichardKarlFuchs committed:
modify zip call

RichardKarlFuchs committed:
add resetting of lanes to laneType through

RichardKarlFuchs committed:
set SubRoute method to be abstract, add SubRoute implementation for Roundabout

RichardKarlFuchs committed:
add method to select Element from IList based on a weight

RichardKarlFuchs committed:
move choice to new class RoadProbability to improve complexity

RichardKarlFuchs committed:
fix pathfinding, add last RouteSegment

RichardKarlFuchs committed:
add probability to choose a route based on array of weights

RichardKarlFuchs committed:
fix selection of next routeSegment for cars and construction of weights

RichardKarlFuchs committed:
add Crossection Constructor, fix subroute bug, fix despawn bug

1NG3M4R commented on Assets/Scripts/DataTypes/Vertices/TeeSection.cs:
```diff
@@ -15,6 +16,37 @@ public TeeSection(GameObject prefab, Edge throughOrRight, Edge throughOrLeft, Ed
+            if(this.edges.Contains(from) && this.edges.Contains(to))
+                if(this._leftOrRight.Equals(from))
+                    if(this._leftOrRight.Equals(to))
+                        throw new System.Exception("From and to are the same Edge");
```
was ist leftOrRight? ist das eindeutig? wenn nein ist das ne Fehlerquelle

rrickfox commented:
LeftOrRight ist eine private Variable, insofern ist es eindeutig

1NG3M4R <notifications@github.com> schrieb am So., 16. Feb. 2020, 15:47:

> *@1NG3M4R* commented on this pull request.
> ------------------------------
>
> In Assets/Scripts/DataTypes/Vertices/TeeSection.cs
> <https://github.com/rrickfox/traffic-sim/pull/83#discussion_r379908559>:
>
> > +                if(this._leftOrRight.Equals(from))
> +                    if(this._leftOrRight.Equals(to))
> +                        throw new System.Exception("From and to are the same Edge");
>
> was ist leftOrRight? ist das eindeutig? wenn nein ist das ne Fehlerquelle
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rrickfox/traffic-sim/pull/83?email_source=notifications&email_token=AHOLDJ26BRXKZNDYDZFSE7LRDFGZHA5CNFSM4KVZQHR2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCVWCF7Q#pullrequestreview-359408382>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHOLDJ2TGRFJ43BVF6XDEELRDFGZHANCNFSM4KVZQHRQ>
> .
>

1NG3M4R reviewed (commented):
kann das Programm mit den LaneTypes Schon ne SubRoute bilden, und wenn ja, wo?
ist das ganze Schon lauffähig?

rrickfox commented:
Es wird eine Dictionary mit Listen von Subroutes erstellt, nämlich `RoutingTable`. Aus dieser Dictionary wählen Endpoints eine Liste von Subroutes aus, die dann der Reihe nach abgefahren werden. Da die Car Ai allerdings noch nicht in der Lage ist Spurwechsel durchzuführen, können wir das SubRouting noch nicht wirklich demonstrieren. Wie du allerdings in Unity sehen kannst, läuft die Simulation. Ansonsten hätte ich den Pull Request auch nicht erstellt...

funketh committed:
shorten ZipThree()

funketh committed:
refactor ZipThree into Utility

funketh committed:
formatting

funketh commented on Assets/Scripts/DataTypes/Vertices/CrossSection.cs:
```diff
@@ -17,6 +18,54 @@ public CrossSection(GameObject prefab, Edge up, Edge right, Edge down, Edge left
             _left = left;
         }
+
+        public CrossSection(Edge up, Edge right, Edge down, Edge left) : this(EMPTY_PREFAB, up, right, down, left) {}
```
sollte nicht jede CrossSection nen prefab brauchen?

RichardKarlFuchs replied:
Im Intersections Branch ja. Vorher nein

funketh committed:
remove redundant this qualifiers

funketh committed:
unnest SubRoute()

funketh committed:
check whether from and to are the same in the beginning

funketh committed:
make conditional statements more clear

- use == instead of .Equals()
- add comments and redundant else statements to emphasize intent

RichardKarlFuchs committed:
rename method to make funtionality clear

funketh committed:
simplify TeeSection.SubRoute() as well

funketh committed:
Merge branch 'feature/sub_pathfinding' of https://github.com/rrickfox/traffic-sim into feature/sub_pathfinding

funketh commented on Assets/Scripts/DataTypes/Vertices/EndPoint.cs:
```diff
@@ -10,31 +10,48 @@ public class EndPoint : Vertex<EndPoint, EndPointBehaviour>
                 car.Dispose();
-                _edge.cars.Remove(car);
+                _edge.other.cars.Remove(car);
             }
```
Warum die Änderung?

RichardKarlFuchs replied:
`_edge` speichert die Autos die von EndPoint wegfahren. Bisher hat das keinen Fehler erzeugt, das wir nur Straßen hatten, die an beiden Enden EndPoints hatten. Wenn man allerdings eine Straße hat, die nur an einem Ende einen EndPoint hat, despawnt der EndPoint die Autos die von Ihm wegfahren.

funketh committed:
replace lane type reset methods with checks in EndPoint constructor

- all incomingLanes already must only be of type Through => implicitly overriding faulty configurations would not be a good idea

funketh committed:
shorten method and remove redundant using statement

funketh committed:
rename Shift to PopAt and make it more flexible

funketh committed:
move ListExtensions into Utility

funketh committed:
use Equals instead of == when comparing sets in EndPoint()

- this compares their values as opposed to their object id (which would never be the same as they are different objects)

funketh commented on Assets/Scripts/DataTypes/Car.cs:
```diff
@@ -1,54 +1,46 @@
+                track.cars.Remove(this);
+                segment = route.Shift();
+                track.cars.Add(this);
+            }
```
commentate this

RichardKarlFuchs replied:
8ba8f36

funketh committed:
simplify Car.positionOnRoad initialization

RichardKarlFuchs committed:
add commenting

RichardKarlFuchs committed:
Merge branch 'feature/sub_pathfinding' of https://github.com/rrickfox/traffic-sim into feature/sub_pathfinding

funketh commented on Assets/Scripts/DataTypes/Vertices/EndPoint.cs:
```diff
@@ -10,31 +11,46 @@ public class EndPoint : Vertex<EndPoint, EndPointBehaviour>
+
+        public void SetWeights()
+        {
+            _routeProbabilities = new RouteProbabilities(_weights, routingTable.Where(kvp => kvp.Value != null).Select(kvp => kvp.Key).ToList());
```
Weiß nicht wie unmögliche Routen implementiert sind (entweder der Value ist null oder es gibt gar nicht erst nen Eintrag unter dem Key, zu dem es keinen Weg gibt).
Müsste mit dem Pathfinding abgesprochen werden (@Shafma).
```suggestion
            _routeProbabilities = new RouteProbabilities(_weights, routingTable.Keys.ToList());
```

RichardKarlFuchs replied:
Ist abgesprochen, so ists richtig

funketh commented on Assets/Scripts/DataTypes/Edges/ITrack.cs:
```diff
@@ -0,0 +1,12 @@
+{
+    public interface ITrack
+    {
+        RoadShape shape { get; }
```
Sollte man denke ich entfernen, ITrack soll ja nur das interface für autos sein, und die interessiert ja das tatsächliche RoadShape Objekt nicht. Wird auch nirgendwo genutzt.

RichardKarlFuchs replied:
Wird noch nicht genutzt, die Strecken auf den Section werden aber ITrack implementieren.

funketh replied:
Ich weiß, das ist nicht mein Punkt.
Ich meine nur, dass der Sinn von ITrack (und interfaces generell) ist, dass Car mit allen Objekten klar kommt, die dieses interface implementieren, weil Car nur die im interface spezifizierten Methoden/Properties nutzt.
Also macht es keinen Sinn shape da drin zu haben, da Car ja nur indirekt mit der shape interagiert (Deshalb ist sowas wie GetAbsolutePosition ja nicht in Car implementiert, sondern Edge bzw. der Klasse, die ITrack implementiert).
RoadShape ist quasi nen implementationsdetail, um das sich Car entsprechend nicht kümmern sollte.

RichardKarlFuchs replied:
37c00bc

funketh committed:
clean up Edge properties and remove redundant originPoint property

funketh committed:
use System to shorten System.Exception() calls to Exception()

funketh commented on Assets/Scripts/DataTypes/RouteProbabilities.cs:
```diff
@@ -0,0 +1,59 @@
+        // modified from: https://stackoverflow.com/a/43345968
+        public IVertex Choose()
+        {
+            var random = new Random();
```
vielleicht nen globales random Objekt nutzen: https://stackoverflow.com/questions/2643421/correct-method-of-a-static-random-next-in-c

RichardKarlFuchs replied:
37c00bc

funketh commented on Assets/Scripts/DataTypes/Vertices/EndPoint.cs:
```diff
@@ -10,31 +11,46 @@ public class EndPoint : Vertex<EndPoint, EndPointBehaviour>
+        public void SetWeights()
+        {
+            _routeProbabilities = new RouteProbabilities(_weights, routingTable.Where(kvp => kvp.Value != null).Select(kvp => kvp.Key).ToList());
         }
```
gefällt mir irgendwie nicht ganz, dass diese Methode in EndPoint drin ist und EndPoint nur funktioniert, wenn es in Pathfinding nicht vergessen wird sie aufzurufen. Gib bestimmt ne bessere Methode.
Außerdem: #84, aber das ist denke ich klar

RichardKarlFuchs replied:
Wir ja nur einmal aufgerufen und dabei werden auch alle aufgerufen. Von daher kein Problem

funketh commented:
Ist aber unschön

On Tue, 18 Feb 2020, 13:36 Richard Fuchs, <notifications@github.com> wrote:

> *@rrickfox* commented on this pull request.
> ------------------------------
>
> In Assets/Scripts/DataTypes/Vertices/EndPoint.cs
> <https://github.com/rrickfox/traffic-sim/pull/83#discussion_r380644781>:
>
> > +        public void SetWeights()
> +        {
> +            _routeProbabilities = new RouteProbabilities(_weights, routingTable.Where(kvp => kvp.Value != null).Select(kvp => kvp.Key).ToList());
>          }
>
> Wir ja nur einmal aufgerufen und dabei werden auch alle aufgerufen. Von
> daher kein Problem
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rrickfox/traffic-sim/pull/83?email_source=notifications&email_token=AK3W7LSBRW6B4MAV3XC3YNDRDPI5JA5CNFSM4KVZQHR2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCV44PJY#discussion_r380644781>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AK3W7LWL6AMX5OR2JAIU6X3RDPI5JANCNFSM4KVZQHRQ>
> .
>

rrickfox commented:
> 
> Ist aber unschön
> […](#)
> On Tue, 18 Feb 2020, 13:36 Richard Fuchs, ***@***.***> wrote: ***@***.**** commented on this pull request. ------------------------------ In Assets/Scripts/DataTypes/Vertices/EndPoint.cs <[#83 (comment)](https://github.com/rrickfox/traffic-sim/pull/83#discussion_r380644781)>: > + public void SetWeights() + { + _routeProbabilities = new RouteProbabilities(_weights, routingTable.Where(kvp => kvp.Value != null).Select(kvp => kvp.Key).ToList()); } Wir ja nur einmal aufgerufen und dabei werden auch alle aufgerufen. Von daher kein Problem — You are receiving this because you were assigned. Reply to this email directly, view it on GitHub <#83?email_source=notifications&email_token=AK3W7LSBRW6B4MAV3XC3YNDRDPI5JA5CNFSM4KVZQHR2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCV44PJY#discussion_r380644781>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AK3W7LWL6AMX5OR2JAIU6X3RDPI5JANCNFSM4KVZQHRQ> .

Also?

RichardKarlFuchs committed:
add global Random Instance, rename Car.SetPosition() to Car.UpdatePosition(), re-add Edge.originPoint, remove RoadShape from ITrack, rename EndPoint.SetWeights() to EndPoint.CalculateRouteProbabilities()
This commit has been sponsored by Raid Shadow Legends

Shafma committed:
Fix check for LaneType in EndPoint constructor

Shafma committed:
Fix comments

Shafma committed:
Cleanup

Shafma commented on Assets/Scripts/Pathfinding/Pathfinding.cs:
```diff
@@ -96,9 +99,20 @@ private static List<Edge> DetermineFoundPath(this EndPoint self, IVertex end)
+                vertexPath.Skip(1),
+                vertexPath.Skip(2),
+                (v1, v2, v3) =>
+                    new RouteSegment(track: v1.GetEdge(v2), laneType: v2.SubRoute(v2.GetEdge(v1), v2.GetEdge(v3)))
```
Wieso wird bei der SubRoute die Edge von v2 zu v1 genutzt und nicht anders herum?

RichardKarlFuchs replied:
Weil in v2 nur die Edge von v2 zu v1 gespeichert ist. Die von v1 nach v2 ist ja ein anderes Objekt.

Shafma replied:
Aber das ändert doch nicht, dass wir die Edge haben wollen, die zur Vertex führt. Und das ist doch v1 zu v2, oder?

RichardKarlFuchs replied:
In subroute geht es bloß um den benötigten LaneType um von der ersten Edge zur zweiten Edge zu kommen. Die Richtung der Edge ist dabei egal, wichtig ist nur, dass die Edges in der Section mit Namen (Lagebeziehung) gespeichert sind. Der eigentliche Track ist `v1.GetEdge(v2)`, also die edge von v1 nach v2.

Shafma replied:
Ja, war nen Denkfehler.
Vielen Dank für die aufschlussreiche Antwort und produktive Diskussion.

funketh replied:
Ist denke ich nen bisschen unintuitiv, dass SubRoute so implementiert ist.
Wäre dafür, dass SubRoute bei seiner ersten Edge erstmal `.other` aufruft.
Wenn man nämlich die SubRoute Funktion sieht, denkt man dass sie einem sagt, wie man abbiegen muss, wenn man auf edge1 steht und zu edge2 möchte.

RichardKarlFuchs replied:
e4c03cd

RichardKarlFuchs committed:
refactor SubRoute to call .other on incoming Edge

Shafma committed:
Add custom exception

funketh reviewed (changes requested):
Vorschlag:
- RouteProbabilites wird zu einer einzigen static methode `Choose(destinations, weights)` in ner static klasse `Random` in `Utility`
    - (Falls einen der performance verlust juckt, da `CalculateCumulative()` quasi unnötig mehrfach berechnet wird, kann man das auch global cachen in nem dictionary von weights zu cumulativeWeights)
- ==> `EndPoint.CalculateRouteProbabilities()` kann komplett weg

funketh reviewed (changes requested):
Vorschlag:
- RouteProbabilites wird zu einer einzigen static methode `Choose(destinations, weights)` in ner static klasse `Random` in `Utility`
    - (Falls einen der performance verlust juckt, da `CalculateCumulative()` quasi unnötig mehrfach berechnet wird, kann man das auch global cachen in nem dictionary von weights zu cumulativeWeights)
- ==> `EndPoint.CalculateRouteProbabilities()` kann komplett weg

RichardKarlFuchs committed:
move calculation of route probabilities to static class Utiltiy.Random

RichardKarlFuchs committed:
fix check for LaneTypes of Edges in EndPoint

RichardKarlFuchs committed:
remove unused crosssection constructor

funketh committed:
use IList instead of List for destinations

funketh committed:
fix typo

funketh reviewed (approved):
Sieht jetzt gut aus
